packages/kobotoolbox/src/util.js

```diff
@@ -0,0 +1,122 @@
+import { composeNextState } from '@openfn/language-common';
+import {
+  request as commonRequest,
+  makeBasicAuthHeader,
+  logResponse,
+} from '@openfn/language-common/util';
+
+export function prepareNextState(state, response) {
+  const { body, ...responseWithoutBody } = response;
+  return {
+    ...composeNextState(state, body),
+    response: responseWithoutBody,
+  };
+}
+export function request(state, method, path, opts = {}) {
+  const { baseUrl, apiVersion, username, password } = state.configuration;
+  const { data = {}, query = {}, headers = {}, parseAs = 'json' } = opts;
+
+  const requestPath = cleanPath(`/api/${apiVersion}/${path}`);
+  const authHeaders = makeBasicAuthHeader(username, password);
+
+  const options = {
+    body: data,
+    headers: {
+      'Content-Type': 'application/json',
+      ...authHeaders,
+      ...headers,
+    },
+    query: {
+      format: 'json',
+      ...query,
+    },
+    parseAs,
+    baseUrl,
+  };
+
+  return commonRequest(method, requestPath, options).then(logResponse);
+}
+
+export async function requestWithPagination(state, path, options = {}) {
+  const results = [];
+
+  let { pageSize, start, limit, max, ...otherOptions } = options;
+
+  const isUsingDefaultMax = max === undefined;
+  const maxResults = max ?? limit ?? 1e4;
+
+  let isFirstRequest = true;
+  let requestOptions = { query: { start, limit }, ...otherOptions };
+  let shouldFetchMoreContent = false;
+  const didUserPassLimit = Boolean(max || limit);
+
+  do {
+    requestOptions.query ??= {};
+
+    if (!isNaN(start)) {
+      requestOptions.query.start = start;
+    }
+
+    if (limit) {
+      requestOptions.query.limit = limit;
+    } else if (didUserPassLimit || !isFirstRequest) {
+      // If there's an explicit limit or page size,
+      // or this is not the first request,
+      // set the limit in the URL
+      requestOptions.query.limit = Math.min(
+        pageSize || maxResults,
+        maxResults - results.length
+      );
+    } else if (!isNaN(pageSize)) {
+      requestOptions.query.limit = pageSize;
+    }
+
+    const response = await request(state, 'GET', path, requestOptions);
+
+    if (response.body.results) {
```

Comment:
This needs to be `response.body?.results`, in case there's no response body (I'm seeing this right now with a 301 response)

Or maybe if there's no body we exit the loop?

Reply:
Fixed